### ISSUE TITLE
Add Socket connection?

### ISSUE THREAD

**Issue by zoakes** [da4089/simplefix#22]
I've been thinking about adding a socket connection class to a generic FIX IP, was curious if there's anything in beta for this?  If not, I'm happy to fork and contribute what I can to it if I have some success -- I was going to connect to IB.

Let me know your thoughts!

Simplefix is INCREDIBLE, so simple -- exactly what I was looking for -- FIX formatting without the overhead mess of QuickFix -- just needs a Socket initiator : )

Zach

**Comment by da4089** [comment]
Hi Zach,

At one time I was fairly opposed to this idea.  I think it's a slippery slope from adding some sort of socket support, and then maybe some sort of message validation, and suddenly we're back where QuickFIX is.

But I'm maybe less against it than I was, mostly because I think a lot of people end up having to write it themselves, and while that's not terribly hard, it does seem like maybe an off-the-shelf way to do it would be better.

So.  This is maybe a little unfair, but I'd kinda be interested in seeing what you come up with, without really promising to merge the PR if I get cold feet.  But ... since you're writing it anyway, and since you're kinda volunteering, I'm absolutely happy to look at a PR and work with you on merging it if it seems promising?

**Comment by zoakes** [comment]
Sure. Well mine is very simple, but I’m happy to send.

I have it on a git, in the Socket+ branch (with generic connection now)

https://github.com/zoakes/sFIX-IB/blob/Socket%2B/SFX_Sock.py



Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by zoakes** [comment]
It’s far from QuickFix haha — I like simple too. QF is just so bloated, and their documentation is terrible.  I hate QF.

Also — I realize I need to remove the finally’s to the very bottom to keep connection open (in sendAll) — just was benchmarking and can’t if it’s running continuously.

Goal was the most minimal connection to IB’s FIX protocol.

Zach

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by zoakes** [comment]
I added some stuff that most people would probably want -- like heartbeat, logging, etc.  I think for now I'm done, this worked for me today in some preliminary testing, so I'm pretty pleased hah.  I expected this fix engine to take like.. weeks-months, so thank you for saving me that : ) 
End goal is Cython /  C++ engine, so I'm going to get started on optimizing tomorrow.

I'm happy to work with you on the PR (pull request?) -- I'm not super familiar with all the github ins and outs, but I'm a solid developer. If it's beginning to feel a bit like QF, no need to merge -- I'm indifferent, really I'm happy I have that makes sense and works well in a few hundred lines.  Let me know !

Thank you again for this incredible module.

Zach

**Comment by zoakes** [comment]
David,

I finished a really clean Python FIX client with simplefix -- I'll attach the current version here. I finally finished up integrating / testing with Interactive Brokers IB CTCI FIX API, they are incredibly particular about what's used, what isn't, what's in header, etc -- but it's ready now!
If you'd like to use it, or pull it --  you're welcome to. 
I currently have it running the heartbeat as a Daemon thread, but I have a few versions of heartbeat that can be used -- this has been the most consistent. 

https://github.com/zoakes/sFIX-IB/blob/master/base_hb_v2.0.py

I didn't add any message validation, I'm not quite sure how to aside from just using a listener -- which sounds bloated.  If it is something you're interested in I can give it a try though.

**Comment by mecanda12** [comment]
@da4089 @zoakes Hey David, 

If Zoakes, code is worth it, is any possibility to merge it? I see his code help connect to the backend engine.

**Comment by mecanda12** [comment]
Thanks, David :)

**Comment by zoakes** [comment]
My code was pretty minimal — more scaffolding to build out your own spec with if I recall.
Honestly, there’s a lot of FIX overhead in maintaining connectivity, and with the new FIX5 / SP stuff (session and transport layer split), it’s even more substantial.

I can run over it again, but I didn’t intend this to be merged into main.
If you find use in it, go for it of course :)

I’m guessing something generic that applies + passes cert at various FIX endpoints would be substantially larger — and not really fit with the idea here.

We run FIX at CME on ILink3 and it’s like thousands of lines of code to maintain connectivity, persist messages for reconnect / disconnect / reconciling msg counts, etc.
it’s not small.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by da4089** [comment]
For better or worse, there's not a lot of usage of the new session layer stuff that I've seen: most FIX is still using the classic session protocol and some variant of FIX4.x.

But yes, a production quality FIX engine is a heap of work.  In most cases, you'd probably be better off starting with the Python QuickFIX or WTFIX if that's what you need.

That said, something that's reasonably resilient and suitable for running regression tests, etc, isn't too much effort.

**Comment by mecanda12** [comment]
@da4089 I am looking for a solution for a Non-production environment. Most a Test environment. 
My question is the same how can I make a network connection using simplefix, QuickFIX, or WTFIX?

**Comment by zoakes** [comment]
Yep. Even then -- quickfix doesn’t handle the new ‘split layer’ ILink3 stuff, or doesn’t publish anything implying they can.

We ended up building one, but it took months.

For basic stuff, sure – could be fairly minimal. Just a heartbeat, some other minimal overhead.
If you want something to truly manage a prod level OMS / MD connection – it will quickly look like QuickFix.  If that was the goal, I’d say this probably deviates from the ‘simple’ goal of the library.

Also of note – many of the paid FIX engine players barely have FIX5 / SP stuff working / tested.
CME is shifting entirely to ILink3 in 2024, phasing out ILink2 now (FIX4), but no new connections can use FIX4 for example.  It’s a bit of a unique time in FIX software where everyone w DMA is preparing to scramble or break.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by zoakes** [comment]
How can you make a network connection… as in how do you handle it from a software perspective?
or how do you get the connectivity?

Software is fairly simple – the connectivity maybe not. DMA will run in the hundreds of thousands. Some brokers offer FIX alternatives to REST stuff, like IB, but that’s sort of FIX – lite.

Just depends, but you’ll likely need to bring some $ to the table for anyone to certify with you.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: Zach ***@***.***>
Sent: Wednesday, April 5, 2023 11:23 AM
To: ***@***.***>; ***@***.***>
Cc: ***@***.***>
Subject: RE: [da4089/simplefix] Add Socket connection? (#22)

Yep. Even then -- quickfix doesn’t handle the new ‘split layer’ ILink3 stuff, or doesn’t publish anything implying they can.

We ended up building one, but it took months.

For basic stuff, sure – could be fairly minimal. Just a heartbeat, some other minimal overhead.
If you want something to truly manage a prod level OMS / MD connection – it will quickly look like QuickFix.  If that was the goal, I’d say this probably deviates from the ‘simple’ goal of the library.

Also of note – many of the paid FIX engine players barely have FIX5 / SP stuff working / tested.
CME is shifting entirely to ILink3 in 2024, phasing out ILink2 now (FIX4), but no new connections can use FIX4 for example.  It’s a bit of a unique time in FIX software where everyone w DMA is preparing to scramble or break.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by mecanda12** [comment]
@zoakes To make a connection with the backend engine, a socket can be used to communicate with the bundle. 

Ip address, Port, Target & Sender comp Id, Target Sub ID, User address and password 

How can you make a network connection… as in how do you handle it from a software perspective?
or how do you get the connectivity?

**Comment by zoakes** [comment]
These requirements are all set per endpoint.
There’s a certification process involved with most? (All?), all I’ve connected to.  Depends on where you are connecting.

I’ll also say the VPN / VPC requirements for many of these aren’t trivial either.

To get access to this whole environment , usually a matter of $$.
There’s minimum volumes, or account sizes, etc,

So like… CME connectivity requires I think $1M account size, 10k contracts/mo, something.
It may also require membership, unsure.
We have membership, unsure if that’s a req.

Easier routes would be broker fix, IB, probably 100k requirement.
TT is another one, but that’s like 3k/mo ish.

It’s really just going to require $ in all the US Equities / Futures endpoints I’m familiar with. Maybe there’s some FX or digital stuff that’s ‘free’ and easier, don’t know.
I recall darwinex had an endpoint but closed it to new users.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mecanda12** [comment]
@zoakes I do not think regression testing needs that much $$. 
Keep in mind I am talking about internal-level testing. Which will be happening on the same network. 

Do you have a solution for that?

**Comment by zoakes** [comment]
Sure, you can mock a FIX endpoint.  Quickfix has one ready to go, though you'll need to hard code various behaviors into it to sort of 'mock' the behavior you expect wrt resets, rejections, sessions, etc. 

To connect to a real endpoint (for any form of real testing), I think you will need to pass certification & fully integrate for the given endpoint.  
That's always been the case for me. 
My SFIX-IB was basically a hacked solution to pass certification with IB, but eventually I used QuickFix to manage the overhead of things like sequence reset, etc. 

Real fix is complicated, and the certifications require handling a lot of edge cases.  
CME's was literally like... send 60,000 limit orders, we'll fill / reject them randomly, then cancel any remaining. 
Now add a disconnect + reset / reconcile halfway through, etc.
IB's was exponentially easier, but it's also very manual whereas cme is automated certification, just a testing suite. 

I don't really see the point of testing the robustness of a mocked connection, personally, but that's just me.  
These multicast, distributed systems are not trivial to test, I find it more efficient to write larger tests that cover as much as possible.  
Again -- that's purely my preference.

**Comment by Umair-A** [comment]
Hello, there! 
As, I'm new to the FIX protocol I do not understand where I can start and how I can use this if zoakes or anyone else can give me any suggestions or documents to understand this it will be very helpful for me the real fix documentation is really overwhelming and going over my head.

**Comment by zoakes** [comment]
Not following... I don't think MT5 has FIX, I don't know why they would. 
Fix would be more for connecting directly to venues -- cme, nyse, Nasdaq, bzx, edga, edgx etc. 

Some brokers provide endpoints that can then route to many venues -- trading technologies, interactive brokers. 

I don't know where a trading platform would come in. 
FIX is just a messaging protocol. 

I see no mention of mt5 offering a fix endpoint, don't see what purpose it would serve their users.

**Comment by Umair-A** [comment]
Exactly, I will get an endpoint from the broker. I wanted to get market feed data and other order sending on it etc. If you could please look into this repository it is what exactly I want but it is an old repository which is not working now.
[https://github.com/rinleit/quickfix-python-samples](url)

**Comment by Umair-A** [comment]
And If you can please give me steps on how can I achieve a connection with any server using quickfix or fix API. Please mention some steps or material if it is possible.

**Comment by zoakes** [comment]
Again — this is not a quickfix issue forum.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
